gfx.cpp

```diff
@@ -8849,6 +8856,86 @@ class GfxInternal
             }
         }
 
+        uint64_t shader_key = 0;
+        std::string shader_key_bytecode;
+        std::string shader_key_reflection;
+        std::vector<WCHAR> wshader_key_bytecode;
+        std::vector<WCHAR> wshader_key_reflection;
+        if constexpr(std::is_same<ID3D12ShaderReflection, REFLECTION_TYPE>::value)
+        {
+            IDxcResult *dxc_preprocess = nullptr;
+            shader_args.push_back(L"-P");   // run DXC as preprocessor
+            dxc_compiler_->Compile(&shader_source, shader_args.data(), (uint32_t)shader_args.size(), dxc_include_handler_, IID_PPV_ARGS(&dxc_preprocess));
+            if(dxc_preprocess != nullptr)
+            {
+                IDxcBlob *dxc_hlsl = nullptr;
+                dxc_preprocess->GetOutput(DXC_OUT_HLSL, IID_PPV_ARGS(&dxc_hlsl), nullptr);
+                if(dxc_hlsl != nullptr)
+                {
+                    std::string const hlsl((char *)dxc_hlsl->GetBufferPointer(), dxc_hlsl->GetBufferSize());
+                    HashCombine(shader_key, Hash(kernel.entry_point_.c_str()));
+                    for(String const &define : kernel.defines_)
+                        HashCombine(shader_key, Hash(define.c_str()));
+                    HashCombine(shader_key, Hash(hlsl.c_str()));
+                    HashCombine(shader_key, shader_type);
+                    dxc_hlsl->Release();
+                }
+                dxc_preprocess->Release();
+            }
+            shader_args.pop_back();
+            if(shader_key != 0)
+            {
+                std::map<uint64_t, Shader>::const_iterator const it = shaders_.find(shader_key);
+                if(it != shaders_.end())
+                {
+                    shader_bytecode = (*it).second.shader_bytecode_;
+                    reflection = (*it).second.shader_reflection_;
+                    GFX_ASSERT(shader_bytecode != nullptr && reflection != nullptr);
+                    if(dxc_source) dxc_source->Release();
+                    return; // done
+                }
+                std::string shader_key_file = "./shader_cache/";
```

Comment:
MiniNN already creates a cache folder called "cache", the file extensions and naming are different so it may be usefull to reuse the same directory to keep things neat

Reply:
Yea, that's a good point, and I thought of this too.
On another hand, it's kinda neat that the 2 caches are separate (HIP vs. D3D12) as they store very distinct artefacts after all?
Not completely sure which I prefer on that one 😛